docs/news.py

```diff
@@ -0,0 +1,164 @@
+
+    print('Generating Markdown news...')
+    run('./build.py', 'markdown')
+    with (docs_path / '_build/markdown/temp_news.md').open('r') as f:
+        this_release_md = f.read()
+    print(this_release_md)
+
+    insert_into_file(docs_path / '../NEWS.md', '# OpenDDS Releases', this_release_md)
+
+    print('Removing fragments...')
+    for fragment in get_fragments():
+        print(fragment)
+        fragment.unlink()
+
+    return 0
+
+
+def get_pr_num(fragment):
+    # Get the last commit hash
+    sha = subprocess.check_output(
+        ['git', 'log', '-n', '1', '--pretty=format:%H', '--', str(fragment)],
+        cwd=str(docs_path)).decode('utf-8')
+    print('  sha', repr(sha))
+
+    # Then try to use that to ask GitHub for the PR with that commit
+    pr_num = subprocess.check_output(
+        [
+            'gh', 'pr', 'list', '--search', sha, '--state', 'merged',
+            '--json', 'number', '--jq', '.[0].number'
+        ],
+        cwd=str(docs_path)).decode('utf-8').strip()
+    print('  pr_num', repr(pr_num))
+
+    return pr_num if pr_num else None
+
+
+def insert_pr_num(args):
+    if 'GITHUB_ACTIONS' not in os.environ:
+        sys.exit('ERROR: Can only run on GitHub Actions')
+
+    inserted = False
+    exit_status = 0
+    for fragment in get_fragments():
+        print('Looking in', fragment, '...')
+        with fragment.open('r') as f:
+            lines = f.readlines()
+        with fragment.open('w') as f:
+            for lineno, line in enumerate(lines):
+                line = line.rstrip()
+                m = news_directive_re.match(line)
+                if m and m.group(1) == 'prs':
+                    print('  Found news-prs on line ', lineno)
+                    if 'this' in m.group(2):
+                        pr_num = get_pr_num(fragment)
+                        if pr_num is None:
+                            print('ERROR: could not get PR from last commit of ' + str(fragment) +
+                                ', not just merged in a PR?', file=sys.stderr)
+                            exit_status = 1
+                        else:
+                            print('  Found "this", replacing with', pr_num)
+                            inserted = True
+                            line = m.group(0).replace('this', pr_num)
+                print(line, file=f)
+        if exit_status != 0:
+            break
+
+    if exit_status != 0:
+        print('ERROR: coudn\'t get a PR number, will not push any changes if there are any',
+            file=sys.stderr)
+    elif inserted:
+        print('Pushing changes...')
+        run('git', 'config', 'user.name', 'github-actions[bot]')
+        run('git', 'config', 'user.email', 'github-actions[bot]@users.noreply.github.com')
+        run('git', 'commit', '--all', '--message',
+            'Insert PR Number into News Fragment(s) [skip actions]')
+        run('git', 'push')
+    else:
+        print('No changes to push')
```

Comment:
Can/should this be moved to the github actions workflow?

Reply:
It could, but then I'd have to get the distinction between no change/change/error out of the script. It could be done with the exist status, but this way just seems simpler.